### DISCUSSION TITLE
Create a mapping for several modes at once.

### DISCUSSION THREAD

**Discussion by GaetanLepage** [nix-community/nixvim#576] [General]
Currently, if you want to create a mapping for several modes, you need to duplicate it:
```nix
maps = {
  normal."<C-p>" = { action = "require 'my-plugin'.do_stuff"; lua = true; silent = true; desc = "Does stuff"; ...};
  insert."<C-p>" = { action = "require 'my-plugin'.do_stuff"; lua = true; silent = true; desc = "Does stuff"; ...};
  terminal."<C-p>" = { action = "require 'my-plugin'.do_stuff"; lua = true; silent = true; desc = "Does stuff"; ...};
};
```
Of course, thanks to the flexibility of the nix language, you can do better. But still.

The `lua` code we generate for creating mappings relies on `vim.keymap.set` which accepts as argument a list of modes.
Hence, in lua, the above mappings would translate to this single line:
```lua
vim.keymap.set({"n", "i", "t"}, "<C-p", require 'my-plugin'.do_stuff, { silent = true, desc = "Does stuff", ...})
```

For me, it could be cool to be able to achieve a similar syntax in nixvim.
For instance:
```nix
maps.custom."<C-p>" = {
  mode = ["n" "i" "t"];
  action = "require 'my-plugin'.do_stuff";
  lua = true;
  silent = true;
  desc = "Does stuff";
  ...
};
```

Here are my questions for you nixvim users:
1) Do you feel like this could be useful ? Do you have such duplicated mappings in your config ?
2) What would be a user-friendly syntax/option name for this ?
- Syntax A) In terms of implementation, the easiest would be `maps.custom.KEY = { mode = MODE(S); action = ACTION; ...}`.
  This `custom` mode is for keymaps which declare their mode(s) (or use the default `""` within their attrs.
- Syntax B) We could also allow users to specify mappings directly in the toplevel `maps` attribute set:
```nix
# Current syntax
maps = {
  normalVisualOp.KEY1 = { MAP1 };
  insert.KEY2 = { MAP2 };

  # Duplicated mapping
  normal.KEY3 = { MAP3 };
  terminal.KEY3 = { MAP3 };
};

# Syntax A
maps = {
  normalVisualOp.KEY1 = { MAP1 };
  insert.KEY2 = { MAP2 };

  custom.KEY3 = {modes = ["n" "t"]; MAP3};
};

# Syntax B
maps = {
  normalVisualOp.KEY1 = { MAP1 };
  insert.KEY2 = { MAP2 };

  KEY3 = {modes = ["n" "t"]; MAP3};
};

 # Also correct with syntax B
maps = {
  KEY1 = { MAP1 }; # mode is omitted because the default is "" which translates to normalVisualOp
  KEY2 = { mode = "i"; MAP2 };

  KEY3 = {modes = ["n" "t"]; MAP3};
};
```

**Comment by traxys** [comment]
I feel that mapping as closely as possible to neovim is the right way to go, so I'm more for something like syntax "B" that seems to go in that direction.

Optimally we should no even use the `<mode>.<key>` keys, but I think leaving them for backwards compatiblitiy (even if eterenally) would be a good thing.

**Comment by fabianpage** [comment]
I would also prefer `B` as i think it's easier to understand.

**Comment by GaetanLepage** [comment]
Dealt with in #605
